drivers/python/rethinkdb/net_twisted.py

```diff
@@ -0,0 +1,360 @@
+    if timeout == 0 or timeout is None:
+        return
+
+    def raisesTimeout():
+        # If the action hasn't been completed before.
+        if not deferToFail.called:
+            # Raise a timeout error. :)
+            deferToFail.errback(RqlTimeoutError)
+
+    return reactor.callLater(timeout, raisesTimeout)
+
+class DatabaseProtocol(Protocol):
+    WAITING_FOR_HANDSHAKE = 0
+    READY = 1
+
+    def __init__(self, factory):
+        self.factory = factory
+        self.state = DatabaseProtocol.WAITING_FOR_HANDSHAKE
+        self._handlers = {
+            DatabaseProtocol.WAITING_FOR_HANDSHAKE: self._handleHandshake,
+            DatabaseProtocol.READY: self._handleResponse
+        }
+
+        self.buf = str()
+        self.buf_expected_length = 0
+        self.buf_token = None
+
+        self.wait_for_handshake = Deferred()
+
+        self._open = True
+
+    def resetBuffer(self):
+        self.buf = ''
+        self.buf_expected_length = 0
+
+    def connectionMade(self):
+        # Send immediately the handshake.
+        self.transport.write(self.factory.handshake_payload)
+        # Defer a timer which will callback when timed out and errback the
+        # wait_for_handshake. Otherwise, it will be cancelled in
+        # handleHandshake.
+        self._timeout_defer = reactor.callLater(self.factory.timeout,
+                self._handleHandshakeTimeout)
+
+    def connectionLost(self, reason):
+        self._open = False
+
+    def _handleHandshakeTimeout(self):
+        # If we are here, we failed to do the handshake before the timeout.
+        # We close the connection and raise an RqlTimeoutError in the
+        # wait_for_handshake deferred.
+        self.transport.loseConnection()
+        self.wait_for_handshake.errback(RqlTimeoutError())
+
+    def _handleHandshake(self, data):
+        try:
+            self.buf += data
+            if self.buf[-1] == b'\0':
+                message = decodeUTF(self.buf[:-1]).split('\n')[0]
+                if message != 'SUCCESS':
+                    # If there is some problem with the handshake, we errback
+                    # our deferred.
+                    self.wait_for_handshake.errback(RqlDriverError('Server'
+                    'dropped connection with message'
+                    '"{msg}"'.format(msg=message)))
+                else:
+                    # We cancel the scheduled timeout.
+                    self._timeout_defer.cancel()
+                    # We callback our wait_for_handshake.
+                    self.wait_for_handshake.callback(None)
+                    # We're now ready to work with real data.
+                    self.state = DatabaseProtocol.READY
+                    self.resetBuffer()
+        except Exception as e:
+            self.wait_for_handshake.errback(e)
+
+    def _handleResponse(self, data):
+        # 1. Read the header, until we read the length of the awaited payload.
+        if self.buf_expected_length == 0:
```

Comment:
Just after, you can see a `if data >= 12`
If we can't get all the bytes of the response header, we save them, and return from the function in order to wait for more of them.